### PR TITLE
[shared] Remove TagTransformer and improve TagWriter

### PR DESCRIPTION
**Pull request by CodeBlanch**
Follow-up to #5476
Relates to #5475

## Changes

* Removes `TagTransformer` as everything is now using `TagWriter`
* Converts JsonStringArray tests to use `TagWriter` and fix bugs that manifested
* Uses `ReadOnlySpan<char>` for `string`s in `TagWriter` to make it possible to bypass allocations when writing `char`s or`string`s which get trimmed due to length limits

## Merge requirement checklist

* [X] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/CONTRIBUTING.md) guidelines followed (license requirements, nullable enabled, static analysis, etc.)
* [X] Unit tests added/updated

### CODE REVIEW

File: test/OpenTelemetry.Tests/Internal/JsonStringArrayTagWriterTests.cs [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+    [InlineData(new object[] { new bool[] { true, false, false, true }, "[true,false,false,true]" })]
+    public void BooleanArray(bool[] data, string expectedValue)
+    {
+        VerifySerialization(data, expectedValue);
+    }
+
+    [Theory]
+    [InlineData(new object[] { new byte[] { }, "[]" })]
+    [InlineData(new object[] { new byte[] { 0 }, "[0]" })]
+    [InlineData(new object[] { new byte[] { byte.MaxValue, byte.MinValue, 4, 13 }, "[255,0,4,13]" })]
```
Comment (vishweshbankwar) [inline test/OpenTelemetry.Tests/Internal/JsonStringArrayTagWriterTests.cs:46]: Is this format defined in spec somewhere?

Comment (CodeBlanch) [reply test/OpenTelemetry.Tests/Internal/JsonStringArrayTagWriterTests.cs:46]: Here: https://github.com/open-telemetry/opentelemetry-specification/blob/58bfe48eabe887545198d66c43f44071b822373f/specification/trace/sdk_exporters/zipkin.md?plain=1#L134-L135

File: src/Shared/TagWriter/TagWriter.cs [hunk 2]
```diff
@@ -105,47 +111,57 @@ internal abstract class TagWriter<TTagState, TArrayState>
...
-            case byte[] byteArray: this.WriteToArray(ref arrayState, byteArray); break;
-            case sbyte[] sbyteArray: this.WriteToArray(ref arrayState, sbyteArray); break;
-            case short[] shortArray: this.WriteToArray(ref arrayState, shortArray); break;
-            case ushort[] ushortArray: this.WriteToArray(ref arrayState, ushortArray); break;
-            case uint[] uintArray: this.WriteToArray(ref arrayState, uintArray); break;
+            case char[] charArray: this.WriteStructToArray(ref arrayState, charArray); break;
+            case string?[] stringArray: this.WriteStringsToArray(ref arrayState, stringArray, tagValueMaxLength); break;
+            case bool[] boolArray: this.WriteStructToArray(ref arrayState, boolArray); break;
+            case byte[] byteArray: this.WriteToArrayCovariant(ref arrayState, byteArray); break;
+            case short[] shortArray: this.WriteToArrayCovariant(ref arrayState, shortArray); break;
```
Comment (rajkumar-rangaraj) [inline src/Shared/TagWriter/TagWriter.cs:154]: Why some of the entries like `ushortArray` checks were removed. Are these types not expected here?

Comment (rajkumar-rangaraj) [reply src/Shared/TagWriter/TagWriter.cs:154]: Looks like it is covered with `WriteToArrayTypeChecked`. Is there a benefit of splitting it out?

**Review by rajkumar-rangaraj — APPROVED** [review]
LGTM

**Review by vishweshbankwar — APPROVED** [review]

File: src/Shared/TagWriter/TagWriter.cs (hunk 2 above)
Comment (CodeBlanch) [reply src/Shared/TagWriter/TagWriter.cs:154]: Kind of an interesting situation. Check out this fiddle: https://dotnetfiddle.net/xLeGFw

Basically runtime treats sbyte[]/byte[], int[]/uint[], short[]/ushort[], and long[]/ulong[] as the same types when it comes to `is`. The way the code was written before we wouldn't hit all the branches, whichever was first for one of those pairs would win and we would end up casting wrong. Like sbyte[] falls into byte[] which means values like -128 ((sbyte)0x80) becomes 128 ((byte)0x80).

**Pull request merged**